h2o-k8s/README.md

```diff
@@ -0,0 +1,103 @@
+# H2O Kubernetes integration
+
+The integration of Kubernetes and H2O is possible via `water.k8s.KubernetesEmbeddedConfigProvider` - to be found
+in this module. This implementation of `EmbeddedConfigProvider` is dynamically loaded on H2O start and remains inactive
+unless H2O is running in a Docker container managed by Kubernetes is detected.  
+
+## Running H2O in K8s - user's guide
+
+H2O Pods deployed on Kubernetes cluster require a 
+[headless service](https://kubernetes.io/docs/concepts/services-networking/service/#headless-services) 
+for H2O Node discovery. The headless service, instead of load-balancing incoming requests to the underlying
+H2O pods, returns a set of adresses of all the underlying pods. It is therefore the responsibility of the K8S
+cluster administrator to set-up the service correctly to cover H2O nodes only.
+
+### Creating the headless service
+First, a headless service must be created on Kubernetes.  
+
+```yaml
+apiVersion: v1
+kind: Service
+metadata:
+  name: h2o-service
+spec:
+  type: ClusterIP
+  clusterIP: None
+  selector:
+    app: h2o-k8s
+  ports:
+  - protocol: TCP
+    port: 54321
+```
+
+The `clusterIP: None` defines the service as headless. The `port: 54321` is the default H2O port. Users and client libraries
+use this port to talk to the H2O cluster.
+
+The `app: h2o-k8s` setting is of **great importance**, as this is the name of the application with H2O pods inside. 
+Please make sure this setting corresponds to the name of H2O deployment name chosen.
+
+### Creating the H2O deployment
+
+It is **strongly recommended** to run H2O as a [Stateful set](https://kubernetes.io/docs/concepts/workloads/controllers/statefulset/)
+on a Kubernetes cluster. Kubernetes assumes all the pods inside the cluster are stateful and does not attempt to restart
+the individual pods on failure. Once a job is triggered on an H2O cluster, the cluster is locked and no additional nodes
+can be added. Therefore, the cluster has to be restarted as a whole if required - which is a perfect fit for a StatefulSet.
+ 
+
+```yaml
+apiVersion: apps/v1
+kind: StatefulSet
+metadata:
+  name: h2o-stateful-set
+  namespace: h2o-statefulset
+spec:
+  serviceName: h2o-service
+  replicas: 3
+  selector:
+    matchLabels:
+      app: h2o-k8s
+  template:
+    metadata:
+      labels:
+        app: h2o-k8s
+    spec:
+      terminationGracePeriodSeconds: 10
+      containers:
+        - name: h2o-k8s
+          image: 'pscheidl/h2o-k8s'
+          ports:
+            - containerPort: 54321
+              protocol: TCP
+          env:
+          - name: H2O_KUBERNETES_SERVICE_DNS
+            value: h2o-service.h2o-statefulset.svc.cluster.local
```

Comment:
QQ: I don't entirely remember how cluster ips work as opposed to loadbalancers but I would be concerned that the above will not always be the expected behavior. In fact I may recommend defining an `ingress` object https://kubernetes.io/docs/concepts/services-networking/ingress/

Reply:
The defined service is not a load balancer. It's a headless service. There is no load balancing done. It only returns internal IPS (cluster-internal) of all underlying pods (H2O pods in this case) and creates related DNS records. This is only used for H2O clustering.

This is desired behavior, because:

1. IPs of all pods in K8S are internal and should not be visible from the outside. This is not even desired. It's only used for clustering.
2. Making H2O visible from the outside is yet another layer in K8S world. It's a separation of concerns and proper layering principles that makes sure exposure to the outer world can be modified/deleted/created independently of internal cluster state. It's a pattern followed in general.

The DNS service address is there only for H2O to make clustering possible. Which should always be "internal process". Ingresses serve for exposure to the outer worlds, outside of K8S cluster.